### DISCUSSION TITLE
Rust: How to disable semver-compatible updates

### DISCUSSION THREAD

**Discussion by msrd0** [renovatebot/renovate#22694] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I want to receive only those dependency updates where the new version isn't semver-compatible with the old version.

I.e. I don't want PRs like this one: https://github.com/msrd0/gotham_restful/pull/365

### What you tried so far.

I read through the documentation of available options but didn't find anything useful.

**Comment by HonkingGoose** [comment]
### Renovate docs

> I read through the documentation of available options but didn't find anything useful.

Did you _read everything_ in the config option docs, or did you quickly _skim_ the docs to see if anything looked relevant? 😉

Here's your reading list:

- [`cargo` manager](https://docs.renovatebot.com/modules/manager/cargo/)
- [`packageRules`](https://docs.renovatebot.com/configuration-options/#packagerules)
- [`description`](https://docs.renovatebot.com/configuration-options/#description) 
- [`matchManagers`](https://docs.renovatebot.com/configuration-options/#matchmanagers)
- [`matchUpdateTypes`](https://docs.renovatebot.com/configuration-options/#matchupdatetypes)
- [`enabled`](https://docs.renovatebot.com/configuration-options/#enabled)

I recommend you read the _whole_ `packageRules` section of the docs, so from `allowedVersions` to `replacementversion`. This way you get a good overview of the custom behavior you can set.

### Only major updates for all Rust cargo crates?

It sounds like you only want breaking updates for all your Rust cargo crates. Like this?

| Version update | Do you want the update? |  Renovate update type |
| --- | --- | --- |
| `1.0.0` to `2.0.0` | Yes | `major` |
| `1.0.0` to `1.1.0` | No | `minor` |
| `1.0.0` to `1.0.1` | No | `patch` |

If you only want `major` crates updates because Renovate updates your crates too often: read our [Noise Reduction docs](https://docs.renovatebot.com/noise-reduction/) first.

### If you really want `major` updates only

> **Warning**
> Ignoring `minor` and `patch` level updates is a bad idea. You're going to miss out on crucial bugfixes or features. A maintainer may release a security update in a `minor` or `patch` update without creating a CVE/vulnerability alert. Instead of ignoring updates, you should [set a `schedule`](https://docs.renovatebot.com/configuration-options/#schedule) for your updates, or use the [Dependency Dashboard Approval workflow](https://docs.renovatebot.com/key-concepts/dashboard/#dependency-dashboard-approval-workflow). Using the config below is at your own risk.

```json
{
  "packageRules": [
    {
      "description": "Disable Renovate for all Cargo crates.",
      "matchManagers": ["cargo"],
      "enabled": false
    },
    {
      "description": "Enable Renovate for major cargo crate updates only. Later packageRules entries override earlier entries.",
      "matchManagers": ["cargo"],
      "matchUpdateTypes": ["major"],
      "enabled": true
    }
  ]
}
```

**Reply by msrd0** [reply]
Well ... is the update from `0.0.1` to `0.0.2` considered a major update? Because that's semver breaking. But `1.0.1` to `1.0.2` is not semver breaking.

> Ignoring minor and patch level updates is a bad idea.

No. Rust libraries usually only contain the Cargo.toml file, not the Cargo.lock file. Doing non-semver-breaking updates is completely useless as they only set the minimal version. cargo by default will use the latest compatible version with all the available bugfixes.

**Reply by rarkins** [reply]
You are confusing the terms "breaking" and "major". `0.0.1` to `0.0.2` may be a _breaking_ update, but that doesn't make it a _major_ update.

I agree with you on the second point though. I believe @Turbo87 has some intentions on this although perhaps we never created an issue to track it?

**Reply by msrd0** [reply]
Yes, that's exactly why I was asking. Because in the documentation I only saw major/minor/patch/etc but no mentions of semver.

**Reply by rarkins** [reply]
So you could write a package rule which says matchCurrentValue >=1.0.0 and matchUpdateTypes=patch,minor and then enabled=false. This would disable non-major updates for anything with version 1.0.0 or above

**Reply by Turbo87** [reply]
I have a roughly related PR (that I totally forgot about 🙈), but that was about auto-merging semver compatible updates :)

**Reply by msrd0** [reply]
Your suggestion @rarkins seems to be working for simple version requirements, but I also have a requirement

```toml
unindent = ">=0.1, <0.3"
```

and in the dependency dashboard issue, renovate says that `fix(deps): update rust crate unindent to >=0.2.1, <0.3` is awaiting schedule. Which I get is a "semver-breaking update" from 0.1 to 0.2.1, except I don't want that as the version range already includes the latest version.

**Reply by rarkins** [reply]
If you want to disable any updating of complex ranges like that, then you can do something like `"matchCurrentValue": "/(<|>)/", "enabled": false`

**Reply by rarkins** [reply]
I'd be interested to know why Renovate is trying to _bump_ that range though, instead of leave it alone

**Reply by msrd0** [reply]
Hm I don't really want to disable all updates, if version 0.3 gets released I'd like to know. But as long as the latest version is compatible with the range, I don't want it changed.

**Reply by rarkins** [reply]
Ideally it should only be widening, not bumping. It could be solved via config, if you're willing to post a reproduction repo

**Reply by msrd0** [reply]
Sure. My repo is https://github.com/msrd0/gotham_restful

**Reply by rarkins** [reply]
Full repos are unfortunately quite hard to debug. I also don't see any >= < constraint in your root Cargo.tom.

**Reply by msrd0** [reply]
Oh sorry I misunderstood. Here's a new repo: https://github.com/msrd0/renovate-range-issue

**Reply by rarkins** [reply]
Fix for `<` ranges in https://github.com/renovatebot/renovate/pull/22790
